pkg/assembler/assembler.go

```diff
@@ -17,43 +17,66 @@ package assembler
 
 type assembler struct{}
 
-// Identifiable implements the ability to retrieve a set of
-// attributes such that a graph query is able to identify a
-// GuacNode or GuacEdge uniquely (or as a GuacHyperNode).
-type Identifiable interface {
-	// Identifiers returns a map of fields and values which
-	// can be used to identify an object in the graph.
-	Identifiers() map[string]interface{}
-}
-
 // GuacNode represents a node in the GUAC graph
 type GuacNode interface {
-	Identifiable
-
 	// Type returns the type of node
 	Type() string
 
 	// Properties returns the list of properties of the node
 	Properties() map[string]interface{}
+
+	// Attributes returns the names of the properties of the node.
+	//
+	// If a string `s` is in the list returned by `Attributes` then it
+	// should also be a key in the map returned by `Properties`.
+	Attributes() []string
+
+	// IdentifiableAttributes returns a list of tuples of property names
+	// that can uniquely specify a GuacNode.
+	//
+	// Any string found in a tuple returned by `IdentifiableAttributes`
+	// must also be returned by `Attributes`.
+	IdentifiableAttributes() [][]string
```

Comment:
What is the thought process around `IdentifiableAttributes() [][]string`?  Wouldn't this also be just a []string?

Reply:
I think the thought here is that you could have a hash or a purl, or something else.. And a document may only refer to one of them, so it does need to have an alternative to do the linking.

Reply:
We could have scenarios where a node can be uniquely identified by multiple sets of attributes. Each set can be a singleton or contain multiple attributes. E.g., I could identify a node by `id` or by the pair `name`, `hash`, so for that node the function should return `{{"id"},{"name","hash"}}`. I'll add a comment describing this.